### DISCUSSION TITLE
Installation Problem

### DISCUSSION THREAD

**Discussion by cesare190** [mosparo/mosparo#205] [Problems & Issues]
Hello, i'm trying to install Mosparo with apache2 in my vps with traditional installation.
When i visit mosparo.domain.tld/setup my browser redirect me to 404 error page.
I'm using Safari with macOS Sonoma 14.4 and Microsoft Edge as an alternative browser.
My VPS is using Ubuntu Server LTS with mariadb and apache2.

i'm following these steps:

1. I've downloaded and unzupped the lastest release from this repo in my /var/www/mywebsite/mosparo/
2. i've enable a vhost with apache2 that have the document root in /var/www/mywebsite/mosparo/public
3. i've created a mariadb database (for me it's equivalent to mysql) to be used by mosparo
4. when i visit mosparo.domain.tld/i can see only a 404 error page for mosparo.domain.tld/setup

This is my vhost:

```
<VirtualHost *:443>
    ServerName mosparo.domain.tld
    ServerAdmin noreply@domain.tld
    DocumentRoot /var/www/mywebsite/mosparo/public
    ErrorLog /var/www/mywebsite/log/mosparo-error.log
    LogLevel info
    TransferLog /var/www/mywebsite/log/mosparo-access.log

    <IfModule pagespeed_module>
        ModPagespeed off
    </IfModule>

    <IfModule mod_headers.c>
        Header set X-XSS-Protection "1; mode=block"
        Header always set X-Frame-Options "SAMEORIGIN"
        Header always set X-Content-Type-Options "nosniff"
        Header always set Referrer-Policy "strict-origin"
        Header always set X-Powered-By ""
        Header always set Server ""
        Header always set Content-Security-Policy "default-src https:"
    </IfModule>

    <IfModule mod_ssl.c>
        SSLEngine on
        SSLProtocol +TLSv1.2 +TLSv1.3
        SSLCipherSuite ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-PO>
        SSLHonorCipherOrder off
        SSLSessionTickets off
        SSLCertificateFile /etc/letsencrypt/live/domain.tld/fullchain.pem
        SSLCertificateKeyFile /etc/letsencrypt/live/domain.tld/privkey.pem
        SSLOptions +StrictRequire
    </IfModule>

    <IfModule mod_http2.c>
        Protocols h2 http/1.1
    </IfModule>

    <Directory "/var/www/mywebsite/mosparo/public">
        Options FollowSymLinks
        AllowOverride All
        Require all granted
    </Directory>

</VirtualHost>
```

There is a vhost also for http that redirect to https version of the website.

Thanks so much and sorry if the question is stupid. I have seen a similar question but it's not the same problem.

**Comment by zepich** [comment]
Hi @cesare190 

Thank you very much for your question. This is not a stupid question.

You did everything correctly - I cannot see any mistake right now.

Have you checked the error log file of Apache2? It sounds like you have some rewrite issues. You should see more about that in the error log file if that's the case. Is mod_rewrite of Apache2 enabled?

Thank you very much for your answers!

Kind regards,

zepich

**Reply by cesare190** [reply]
Hi, thanks for replying! I'm sure that mod_rewrite is installed and working because i use the same vps for other tools that require the same module, wordpress it's an example case. My log is empty or it shows me only that the page does not exist: 

>  AH00128: File does not exist: /var/www/mywebsite/mosparo/public/setup

Versions:
```
Ubuntu: 22.04.4 LTS
Php Version: 8.3.4
Apache Version: Apache/2.4.58 (Ubuntu)
```

these are the module actually available and installed in apache2:

```
Loaded Modules:
 core_module (static)
 so_module (static)
 watchdog_module (static)
 http_module (static)
 log_config_module (static)
 logio_module (static)
 version_module (static)
 unixd_module (static)
 access_compat_module (shared)
 actions_module (shared)
 alias_module (shared)
 auth_basic_module (shared)
 authn_core_module (shared)
 authn_file_module (shared)
 authz_core_module (shared)
 authz_host_module (shared)
 authz_user_module (shared)
 autoindex_module (shared)
 brotli_module (shared)
 deflate_module (shared)
 dir_module (shared)
 env_module (shared)
 expires_module (shared)
 ext_filter_module (shared)
 fastcgi_module (shared)
 fcgid_module (shared)
 filter_module (shared)
 geoip_module (shared)
 headers_module (shared)
 http2_module (shared)
 mime_module (shared)
 mpm_event_module (shared)
 negotiation_module (shared)
 proxy_module (shared)
 proxy_ajp_module (shared)
 proxy_balancer_module (shared)
 proxy_connect_module (shared)
 proxy_fcgi_module (shared)
 proxy_html_module (shared)
 proxy_http_module (shared)
 proxy_http2_module (shared)
 reqtimeout_module (shared)
 rewrite_module (shared)
 rpaf_module (shared)
 setenvif_module (shared)
 slotmem_shm_module (shared)
 socache_shmcb_module (shared)
 ssl_module (shared)
 status_module (shared)
 unique_id_module (shared)
 xml2enc_module (shared)
```

Permission on folder and file:
- folders and files are owned by www-data:www-data (apache user)
- Folders are in 775, files in 664
- internal files are in 664, folders in 775

**Reply by zepich** [reply]
Hi @cesare190 

Thank you very much for your response.

Very interesting...

Can you access the static files of mosparo, like `https://<your-domain>/build/mosparo.css`?

Or do you get a 404 error for these files, too?

Kind regards,

zepich

**Reply by cesare190** [reply]
Yes it works!

**Reply by zepich** [reply]
Hi @cesare190 

Okay. Can you check that the `.htaccess` file exists in the `/public` directory (`/public/.htaccess`)?

Kind regards,

zepich

**Reply by cesare190** [reply]
Yes .htaccess is in the public directory

**Reply by zepich** [reply]
Hi @cesare190 

Thank you for the update.

What happens if you open the `index.php` in your browser?

`https://<your-domain>/index.php`

Kind regards,

zepich

**Reply by cesare190** [reply]
It shows me the setup page!

**Reply by zepich** [reply]
Do you have a `DirectoryIndex` directive in your Apache2 config somewhere? https://httpd.apache.org/docs/2.4/mod/mod_dir.html#directoryindex

If not, can you add it with `index.php` as value (`DirectoryIndex index.php`) and try it again if it works without the `/index.php` in the URL?

**Reply by cesare190** [reply]
> Do you have a `DirectoryIndex` directive in your Apache2 config somewhere? https://httpd.apache.org/docs/2.4/mod/mod_dir.html#directoryindex
> 
> 
> 
> If not, can you add it with `index.php` as value (`DirectoryIndex index.php`) and try it again if it works without the `/index.php` in the URL?

I'll try, but some days ago I've tried with it and it didn't worked.

**Reply by zepich** [reply]
Hi @cesare190 

Thank you very much for your answer.

Can you test something, please? Please edit the `.htaccess` file in the `/public` directory and add non-working stuff like 'asdfasdf' in the first or last line (not in the `<IfModule>` part). After saving the `.htaccess` file, can you still access mosparo with the `/index.php/` URL or by accessing the CSS file (`https://<your-domain>/build/mosparo.css`)?

If yes, the `.htaccess` file in the `/public` directory is not parsed by Apache2 (because this non-working stuff should throw an error). We have other issues if you see an error (403 or 500).

Kind regards,

zepich

**Reply by cesare190** [reply]
I've tried. With "DirectoryIndex index.php" in vhost or in .htaccess nothing changes.

This is the code of .htaccess:

> DirectoryIndex index.php
> 
> <IfModule mod_rewrite.c>
>     RewriteEngine On
> 
>     RewriteBase /
> 
>     RewriteCond %{HTTP:Authorization} ^(.*)
>     RewriteRule .* - [e=HTTP_AUTHORIZATION:%1]
> 
>     RewriteCond %{REQUEST_FILENAME} !-f
>     RewriteCond %{REQUEST_FILENAME} !-d
>     RewriteRule ^(.*)$ index.php [QSA,L]
> </IfModule>

This is the code of vhost (i'm changing headers for a CORS error in form validation on Wordpress):

```
<VirtualHost *:443>
    ServerName mosparo.domain.tld
    ServerAdmin noreply@domain.tld
    DocumentRoot /var/www/mywebsite/mosparo/public
    ErrorLog /var/www/mywebsite/log/mosparo-error.log
    LogLevel info
    TransferLog /var/www/mywebsite/log/mosparo-access.log

    <IfModule pagespeed_module>
        ModPagespeed off
    </IfModule>

    <IfModule mod_headers.c>
        Header set X-XSS-Protection "1; mode=block"
        #Header always set X-Frame-Options "SAMEORIGIN"
        #Header always set X-Content-Type-Options "nosniff"
        #Header always set Referrer-Policy "strict-origin"
        Header always set X-Powered-By ""
        Header always set Server ""
        #Header always set Permissions-Policy "accelerometer=(), ambient-light-sensor=(), autoplay=(self), battery=(), camera=(), cross-origin-isolated=(self), display-capture=(), document-domain=(), e>
        Header always set Content-Security-Policy "default-src https:"
        Header always set Access-Control-Allow-Origin: "wordpressdomain.tld"
    </IfModule>

    <IfModule mod_ssl.c>
        SSLEngine on
        SSLProtocol +TLSv1.2 +TLSv1.3
        SSLCipherSuite ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-PO>
        SSLHonorCipherOrder off
        SSLSessionTickets off
        SSLCertificateFile /etc/letsencrypt/live/domain.tld/fullchain.pem
        SSLCertificateKeyFile /etc/letsencrypt/live/domain.tld/privkey.pem
        SSLOptions +StrictRequire
    </IfModule>

    <IfModule mod_http2.c>
        Protocols h2 http/1.1
    </IfModule>

    <Directory "/var/www/mywebsite/mosparo/public">
        DirectoryIndex index.php
        Options FollowSymLinks
        AllowOverride All
        Require all granted
    </Directory>

</VirtualHost>
```

**Reply by cesare190** [reply]
Solved with adding directory for /var/www/mywebsite/mosparo/ in apache2.conf
don't know why other websites in mywebsite directory are respecting .htaccess and mosparo/public does not.

**Reply by zepich** [reply]
Okay, very interesting. I'm happy that it's working now! Thank you very much for the update!
